pkg/k8s/watchers/watcher.go

```diff
@@ -93,13 +93,27 @@ func init() {
 		k8s.K8sErrorHandler,
 	}
 
-	k8s_metrics.Register(k8s_metrics.RegisterOpts{
+	// The client-go Register function can be called only once,
+	// but controller-runtime also calls it in its init function
+	// because there's an indirect dependency on controller-runtime.
+	// The metrics of one who called the init first will take effect,
+	// and which one wins depends on the order of package initialization.
+	// Given the possibility that controller-runtime wins, we should set
+	// the adapters directly to override the metrics registration of
+	// controller-runtime as well as calling Register function.
+	// Without calling Register function here, controller-runtime will have
+	// a chance to override our metrics registration.
+	registerOps := k8s_metrics.RegisterOpts{
 		ClientCertExpiry:      nil,
 		ClientCertRotationAge: nil,
 		RequestLatency:        &requestLatencyAdapter{},
```

Comment:
do we still need these if we're just overriding them directly right afterwards?

Reply:
Yes, we still need to call `Register`, otherwise controller-runtime will override our requestResultAdapter.
https://github.com/kubernetes-sigs/controller-runtime/blob/7f0c6dc440f334849b25866dd9990ac1a5a777c2/pkg/metrics/client_go_adapter.go#L52-L54

We need to block it by calling `Register` here.

Reply:
I think I'm still a bit confused, my understanding is that the problem is caused by two places racing to call Register.

Is it not possible for the client_go_adapter callsite to call register prior to this?

Reply:
> Is it not possible for the client_go_adapter callsite to call register prior to this?

It's possible if the import order changes.

client-go `Register` is called from these two init functions, and which will be called first depends on the package initialization order (import order).

`init` in `sigs.k8s.io/controller-runtime/pkg/metrics` 
https://github.com/kubernetes-sigs/controller-runtime/blob/7f0c6dc440f334849b25866dd9990ac1a5a777c2/pkg/metrics/client_go_adapter.go#L43

`init` in `github.com/cilium/cilium/pkg/k8s/watchers`
https://github.com/cilium/cilium/blob/fded1abcf310160a9c518b6e25f1ac6059e08a97/pkg/k8s/watchers/watcher.go#L96

Reply:
@ysksuzuki sorry for the delay, I missed your notification 😞 

What I was getting at is if we look at the [Register code](https://github.com/cilium/cilium/blob/main/vendor/k8s.io/client-go/tools/metrics/metrics.go#L132), as you mentioned this is only done once and the register logic is called inside a sync.Once.

So if this is called after the go-client init one, the Register will still be a no-op which what I think the issue we're trying to fix?

i.e. It seems like this still only works if the pkgs/watchers/watchers call to Registers "wins" and calls Register first

Reply:
> So if this is called after the go-client init one, the Register will still be a no-op which what I think the issue we're trying to fix?

Yes. In the case the controller-runtime wins, we need to override the adapters directly. On the other hand, in the case we win, we need to call `Register`, otherwise controller-runtime will call `Register` and override our adapters. (The former case happens if the import order changes)
This code works for both scenarios.

Reply:
Ok I see, sorry I was a bit confused still.  The metrics are also exported which lets us override them, so in the case where the client-go library sets them we override, otherwise the client-go library just uses Register which is a no-op.

Thanks for the extended clarification 👍

Reply:
Yeah, it's very confusing.  I'm considering whether we can remove the dependency on the controller-runtime's metrics package.